File: app/src/main/java/did/delta/DBHelper.java
package did.delta;

import android.content.ContentValues;
import android.content.Context;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.os.Build;
import android.support.annotation.RequiresApi;
import android.widget.Toast;

import java.io.BufferedReader;
import java.io.InputStreamReader;

public class DBHelper extends SQLiteOpenHelper {

    static public String WORDS_TABLE_NAME = "words";
    static public String SAVE_TABLE_NAME = "save";
    private Context mContext;

    public DBHelper(Context context) {
        super(context, context.getString(R.string.db_name), null, 1);
        mContext = context;
    }

    public DBHelper(Context context, int version) {
        super(context, context.getString(R.string.db_name), null, version);
        mContext = context;
    }

    @RequiresApi(api = Build.VERSION_CODES.ICE_CREAM_SANDWICH)
    @Override
    public void onCreate(SQLiteDatabase db) {
        db.execSQL("create table " + WORDS_TABLE_NAME + " (word varchar(4), description varchar(255), accepted integer);");
        db.execSQL("create table " + SAVE_TABLE_NAME + " (word varchar(4), info varchar(4));");

        try {
            BufferedReader fileReader = new BufferedReader(new InputStreamReader(mContext.getResources().openRawResource(R.raw.words)));
            String[] temwords;
            String temword;

            for (int i = 0; (temword = fileReader.readLine()) != null; ++i) {
                temwords = temword.split(" ");
                ContentValues cv = new ContentValues();
                cv.put("word", temwords[0]);
                cv.put("description", temwords.length > 1 ? temwords[1] : "null");
                cv.put("accepted", temwords.length > 2 ? temwords[2] : "1");
                db.insertOrThrow(WORDS_TABLE_NAME, null, cv);
            }
            fileReader.close();
        } catch (Exception e) {
            Toast.makeText(mContext, e.getLocalizedMessage(), Toast.LENGTH_SHORT).show();
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.ICE_CREAM_SANDWICH)
    @Override
    public void onUpgrade(SQLiteDatabase db, int i, int i1) {
        db.execSQL("drop table if exists " + WORDS_TABLE_NAME + ';');
        db.execSQL("drop table if exists " + SAVE_TABLE_NAME + ';');
        onCreate(db);
    }
}

File: app/src/main/java/did/delta/util/DBHelper.kt
package did.delta.util

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import android.util.Log
import java.io.InputStreamReader

import did.delta.R
import did.delta.bases.WordsListRow

class DBHelper(private val context: Context, version: Int) :
        SQLiteOpenHelper(context, context.getString(R.string.db_name), null, version) {

    fun contains(tableName: String, word: String) : Boolean {
        val c = readableDatabase
                .rawQuery(context.getString(R.string.cmd_count_entrys, tableName, word),
                        null)
        val res = c.moveToFirst() && c.getInt(0) > 0
        c.close()
        close()
        return res
    }

    fun delete(tableName: String, column: String, value: String) {
        writableDatabase.delete(tableName, "$column=?", arrayOf(value))
        close()
    }

    fun setInt(tableName: String, cmpcolumn: String, cmpval: String, column: String, value: Int) {
        try {
            val cv = ContentValues()
            cv.put(column, value)
            writableDatabase.update(tableName, cv, "$column = ?", arrayOf(cmpval))
        } finally {
            close()
        }
    }

    fun rewrite(tableName: String, columns: Array<Map<String, String>>) {
        try {
            val wd = writableDatabase
            wd.execSQL(context.getString(R.string.cmd_delete, tableName))
            wd.execSQL("vacuum")

            columns.forEach { column ->
                val cv = ContentValues()
                column.forEach { (k, v) -> cv.put(k, v) }
                wd.insertOrThrow(tableName, null, cv)
            }
        } finally {
            close()
        }
    }

    fun getInt(tableName: String, cmpcolumn: String, cmpval: String,
               rescolumn: String, default: Int = -1) : Int {
        val c = readableDatabase.
                rawQuery(context.getString(R.string.cmd_usr_query, tableName,
                        cmpcolumn, cmpval, rescolumn),
                        null)
        var res = default
        if (c.moveToFirst()) res = c.getInt(c.getColumnIndex(rescolumn))
        c.close()
        close()
        return res
    }

    fun empty(tableName: String): Boolean {
        val c = readableDatabase.
                rawQuery(context.getString(R.string.cmd_count, tableName), null)
        val res = c.moveToFirst() && c.getInt(0) != 0
        c.close()
        close()
        return res
    }

    fun add(tableName: String, vararg column: Pair<String, String>) {
        val cv = ContentValues()
        column.forEach { (k, v) -> cv.put(k, v) }
        try {
            writableDatabase.insertOrThrow(tableName, null, cv)
        } finally {
            close()
        }
    }

    fun readAll(tableName: String): ArrayList<WordsListRow> {
        val res = ArrayList<WordsListRow>()
        val c = readableDatabase.rawQuery(context.getString(R.string.cmd_full_query, tableName), null)
        if(c.moveToFirst()) do res.add(WordsListRow(c.getString(c.getColumnIndex("word")),
                c.getString(c.getColumnIndex("description")),
                c.getString(c.getColumnIndex("results")))) while(c.moveToNext())
        c.close()
        close()
        return res
    }

    fun readAll2(tableName: String): ArrayList<Pair<String, String>> {
        val res = ArrayList<Pair<String, String>>()
        val c = readableDatabase.rawQuery(context.getString(R.string.cmd_full_query, tableName), null)
        if(c.moveToFirst()) do res.add(c.getString(c.getColumnIndex("word")) to
                c.getString(c.getColumnIndex("description"))) while(c.moveToNext())
        c.close()
        close()
        return res
    }

    fun getRandomWord(tableName: String) : String {
        val c = readableDatabase.rawQuery(context.getString(R.string.cmd_get_random, tableName), null)
        var res: String? = null
        if(c.moveToFirst()) res = c.getString(c.getColumnIndex("word"))!!
        c.close()
        close()
        return res ?: throw Exception("Database is empty!")
    }

    constructor(context: Context) : this(context, 1)

    val gameTypes: Array<out String> = context.resources.getStringArray(R.array.game_types)!!

    override fun onCreate(db: SQLiteDatabase) {
        try {
            db.execSQL(context.getString(R.string.cmd_create_table_highscores,
                    context.getString(R.string.highscores_table_name)))

            gameTypes.forEach {
                game_type ->
                val cv = ContentValues()
                cv.put("game_type", game_type)
                cv.put("score", -1)
                db.insertOrThrow(context.getString(R.string.highscores_table_name),
                        null, cv)

                context.resources.getStringArray(R.array.table_types).forEach {
                    table_type ->
                    db.execSQL(context.getString(R.string.cmd_create_table, table_type, game_type))
                }

                db.execSQL(context.getString(R.string.cmd_create_save_table, game_type))

                val tableName = "words_$game_type"
                val fileReader = InputStreamReader(context.assets.open("$tableName.txt"))
                fileReader.forEachLine {
                    val row = it.split('|')
                    val contentValues = ContentValues()
                    contentValues.put("word", row[0])
                    contentValues.put("description", row.getOrElse(1, { "" }))
                    db.insertOrThrow(tableName, null, contentValues)
                }
            }
        } catch (e: Exception) {
//            context.toast(e.localizedMessage)
            Log.e("did.delta.util.DBHelper", e.localizedMessage)
        }
    }

    override fun onUpgrade(db: SQLiteDatabase, i: Int, i1: Int) {
        gameTypes.forEach {
            db.execSQL(context.getString(R.string.cmd_delete_table, "words_$it"))
        }
        onCreate(db)
    }
}
